### PR TITLE
SLU recipe with wav2vec

### PR DESCRIPTION
**Pull request by Moumeneb1**
Hi, @aheba, @lorenlugosch , @mravanelli , 
As discussed on the slack channel, this is the recipe on timers and such using Wav2Vec fine tuned on ASR as an encoder.

### CODE REVIEW

**Comment by aheba** [comment]
Hey @Moumeneb1 , thanks for this PR, let me run an exp to see if I got the same result

File: recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+# Tokens: BPE with unigram
+# losses: NLL
+# Training: Timers and Such
+# Authors:  Abdelmoumene Boumadane, AbdelWahab Heba, Loren Lugosch 2021
+# ############################################################################
+
+# Seed needs to be set at top of yaml, before objects with parameters are made
+seed: 1986
+__set_seed: !apply:torch.manual_seed [!ref <seed>]
+experiment: wave2vec_classification_seq2seq
```
Comment (aheba) [inline recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml:14]: wav2vec*, why you call it classification_seq2seq ?
maybe: `wav2vec-train-real`

File: recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+number_of_epochs: 50
+batch_size: 4
+lr: 0.0004
+lr_wav2vec: 0.00001
+
+#freeze all wav2vec
+freeze_wav2vec: False
+
+#set to true to freeze the CONV part of the wav2vec model
+freeze_wav2vec_conv: True
```
Comment (aheba) [inline recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml:48]: this maybe added into the wav2vec lobe.. !

File: recipes/timers-and-such/direct/train_with_wav2vec.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+    # We download and pretrain the tokenizer
+    run_on_main(hparams["pretrainer"].collect_files)
+    hparams["pretrainer"].load_collected(device=run_opts["device"])
+
+    hparams["wav2vec2"] = hparams["wav2vec2"].to("cuda:0")
+
+    # freeze the feature extractor part when unfreezing
+    if not hparams["freeze_wav2vec"] and hparams["freeze_wav2vec_conv"]:
+        hparams["wav2vec2"].model.feature_extractor._freeze_parameters()
```
Comment (aheba) [inline recipes/timers-and-such/direct/train_with_wav2vec.py:382]: I think we should move this into the wav2vec lobe..
@TParcollet , what do you think ?
It appear that freezing the conv and fine tuning the Transformer part of wav2vec gives the best results on the SLU..

File: recipes/timers-and-such/direct/train_with_wav2vec.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+            tokens_eos = torch.cat([tokens_eos, tokens_eos], dim=0)
+            tokens_eos_lens = torch.cat(
+                [tokens_eos_lens, tokens_eos_lens], dim=0
+            )
+
+        loss_seq = self.hparams.seq_cost(
+            p_seq, tokens_eos, length=tokens_eos_lens
+        )
+
+        # (No ctc loss)
```
Comment (aheba) [inline recipes/timers-and-such/direct/train_with_wav2vec.py:95]: Remove this comment !

File: recipes/timers-and-such/direct/train_with_wav2vec.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+            print(batch.semantics[0])
+            print(batch.semantics[1])
+
+            for i in range(len(target_semantics)):
+                print(" ".join(predicted_semantics[i]))
+                print(" ".join(target_semantics[i]))
+                print("")
+
+            # craeate json lists
```
Comment (aheba) [inline recipes/timers-and-such/direct/train_with_wav2vec.py:117]: Remove this comment!

File: recipes/timers-and-such/direct/train_with_wav2vec.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+        test_synth_set,
+        all_real_set,
+        tokenizer,
+    ) = dataio_prepare(hparams)
+
+    # We download and pretrain the tokenizer
+    run_on_main(hparams["pretrainer"].collect_files)
+    hparams["pretrainer"].load_collected(device=run_opts["device"])
+
+    hparams["wav2vec2"] = hparams["wav2vec2"].to("cuda:0")
```
Comment (aheba) [inline recipes/timers-and-such/direct/train_with_wav2vec.py:378]: this is not DDP compliant, you need to use `run_opts["device"]`

**Review by aheba — CHANGES_REQUESTED** [review]

File: recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml (hunk 1 above)
Comment (Moumeneb1) [reply recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml:14]: Yes I'll change it

File: recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml (hunk 2 above)
Comment (Moumeneb1) [reply recipes/timers-and-such/direct/hparams/train_with_wav2vec.yaml:48]: I did it on local i'll push on a pull request, so you can choose to freeze the features extractor when instantiating a wav2vec model which is more convenient

File: recipes/timers-and-such/direct/train_with_wav2vec.py (hunk 6 above)
Comment (Moumeneb1) [reply recipes/timers-and-such/direct/train_with_wav2vec.py:378]: the Wave2vec class in speechbrain is not  a sb.Brain class, it's a pytorch model. I'm not sure how to use run_opts on it

**Comment by lorenlugosch** [comment]
I was able to run it, it worked for me.

Some issues:

- Your output folder is called "results_SLU"---the other recipes just use "results"
- The subfolder is called "classification"---this is not a classification problem
- Some typos in the docs ("wave2vec", should be "wav2vec 2.0")
- You finetune wav2vec on ASR, right? If so, you should write somewhere that you're using an ASR version of wav2vec, not the original unsupervised one

**Comment by aheba** [comment]
LGTM for me, I'm able to reproduce the perf..
maybe @mravanelli can do a final review

**Comment by lorenlugosch** [comment]
OK, looks good to me!

(I changed a couple things to make clear that we're still using ASR pre-training and fixed typos)

**Review by aheba — APPROVED** [review]

**Comment by lorenlugosch** [comment]
I can't access my GPU server right now. Can someone else please fix the linter errors (black, yaml, etc)?

**Comment by aheba** [comment]
@Moumeneb1, thank for this PR :) great perf

**Comment by mravanelli** [comment]
I think you should update the README file with the new results..

**Comment by mravanelli** [comment]
As mentioned also by @lorenlugosch, one thing that it is not clear is if this good performance is due to wav2vect or to just due to a big transformer added in the pipeline. Do you have some numbers with a randomly initialized wav2vect model?

**Comment by lorenlugosch** [comment]
A randomly initialized wav2vec model probably won't work well because the version initialized with unsupervised learning (which we use in the ASR recipes) did not work.

**Comment by TParcollet** [comment]
@lorenlugosch please try again, I changed the way weights are initialized.

**Comment by lorenlugosch** [comment]
I don't have access to GPUs right now, but what do you mean? There's only one set of initial weights for the unsupervised wav2vec model, right?

**Comment by aheba** [comment]
I'm uploading the model into the drive..

@TParcollet, @mravanelli  , I have some GPUs avail now, which exp do you want to try ? with which transformer, the ASR transformer trained on librispeech ?

**Comment by lorenlugosch** [comment]
(To be clear, I don't plan to run any more experiments until the end of the summer after my internship. I'm fine with someone else trying more stuff with wav2vec if they want, but I won't)

**Comment by lorenlugosch** [comment]
Actually the README has a problem: the table results are for `train-real + synth`, but your results are for `train-real` only

**Comment by lorenlugosch** [comment]
- [ ] Add results for wav2vec trained on `train-real + synth`
- [ ] Add results for baseline direct model trained with batch size 8

**Comment by mravanelli** [comment]
I ran the full training and I got the following results:

```
Epoch loaded: 38 - test loss: 7.11e-01, test CER: 2.85e-01, test WER: 1.11, test SER: 7.92 (test-real)
Epoch loaded: 38 - test loss: 7.23e-01, test CER: 1.06, test WER: 3.23, test SER: 18.87 (test_synth)
```

They look similar to @aheba's results (at least within the expected range of variability). Feel free to use my run as well to estimate the standard deviation of the performance (see README). One weird thing is that the performance on `test_synth` is way worse than that with our previous system. I tend to trust more real-data, but the performance difference look important. I don't know if @lorenlugosch has some more comments on that. 
Beyond addressing @lorenlugosch's comments above, I would ask you to:
1- add the model on google drive
2- add the model on HuggingFace

**Comment by aheba** [comment]
I just shared the model in the GDrive:
https://drive.google.com/drive/folders/1s386SAfIVkAcvEGGKmKJ1GD-IDhucbxm?usp=sharing

@mravanelli, notice that the model are trained only on `train-real`. With the CRDNN encoder, we have a little bit worse perf (from: https://drive.google.com/drive/folders/1TcVQwCk5rHJXkcin-XUKAhhUniD6aWSR?usp=sharing) :
```
Epoch loaded: 49 - test loss: 6.69e-01, test CER: 3.45e-01, test WER: 1.90, test SER: 15.00
Epoch loaded: 49 - test loss: 7.50e-01, test CER: 1.38, test WER: 4.41, test SER: 24.82
Epoch loaded: 49 - test loss: 6.69e-01, test CER: 3.35e-01, test WER: 1.95, test SER: 15.42
Epoch loaded: 49 - test loss: 6.69e-01, test CER: 3.35e-01, test WER: 1.95, test SER: 15.42
Epoch loaded: 49 - test loss: 6.69e-01, test CER: 3.35e-01, test WER: 1.95, test SER: 15.42
Epoch loaded: 49 - test loss: 6.69e-01, test CER: 3.35e-01, test WER: 1.95, test SER: 15.42
Epoch loaded: 49 - test loss: 7.49e-01, test CER: 1.38, test WER: 4.41, test SER: 24.89
```
SER of 24.89 compared to 18,87.. !

Let me share the model on HuggingFace..

@Moumeneb1 , could you please share with us the results on `train-real-synth` and the exp with a batch of 8. As far as I remember you are not able to train with batch=8 on the GPU V100 16gb ? could you try to run the exp on the RTX6000 24Gb server?

**Comment by Moumeneb1** [comment]
@aheba I'm still running the experiments on real-synth, on the recipe bash file, you ran them for two epochs, right? And yes, we were not able to run 8 batches on the 16gb when some parts of wav2vec thawed, The experiments are running on the RTX GPUs now, I'll let you know once the experiments are finished.

@lorenlugosch The basic wav2vec model did not work well for us either. When we didn't freeze it, we got an  SER of 43.084 +- 5.61 compared to the wav2vec2 model adapted to ASR where we got 22.52+-4.62. I have started a new experiment of the wav2vec base with transformers finetunning, I will come back to you with the exact results. I also think that the finetunning of wav2vec on the ASR shifts the weights of the model a lot from it's main pre-training objectif. A similar thing was observed on this paper https://arxiv.org/pdf/2104.03502.pdf.

**Comment by Moumeneb1** [comment]
@lorenlugosch  @aheba  I got better results with a batch of 8 an SER of 5.99 +- 1.16 on the finetuned model on ASR,  and 7.334 +-1.00 SER on the wav2vec2 base ( the one not fine tuned os ASR) when  unfreezing the transformers and keeping the CNN frozen.  Apparently the partial unfreezing of the wav2vec can compensate for the finetunning on ASR.

**Comment by TParcollet** [comment]
What is the current state of this PR, are we waiting for an updated recipe ?

**Comment by lorenlugosch** [comment]
Nice!

If you add your scores for these additional models to the README, I can do a last check before we merge this. (unless you have more experiments you want to run)

**Comment by aheba** [comment]
We don't have enough GPU for training the train-synth and train-all.. If you are ok with the results and the PR, let Merge it..

I will report the results and the model on Gdrive.

Here is the train-real with `batch=8`, the best model got 4.8 SER:
https://drive.google.com/drive/folders/1yPhnXm1HbrQyEp_C0XeyFwwFsrkfHXP8?usp=sharing

**Comment by aheba** [comment]
@lorenlugosch , if you are ok, LTGM

**Comment by lorenlugosch** [comment]
I added a new table with the improved results, including @Moumeneb1's unsupervised version.

Great work!

**Pull request merged**